### PR TITLE
Feat(ToAvro):new UDF to write schema as Avro to the Schema Registry

### PR DESCRIPTION
**Pull request by Doe-Ed**

### CODE REVIEW

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+          SchemaManager.PARAM_VALUE_SCHEMA_NAMING_STRATEGY               -> SchemaManager.SchemaStorageNamingStrategies.TOPIC_RECORD_NAME,
+          SchemaManager.PARAM_VALUE_SCHEMA_NAME_FOR_RECORD_STRATEGY      -> parameters.get("value.schema.name").get,
+          SchemaManager.PARAM_VALUE_SCHEMA_NAMESPACE_FOR_RECORD_STRATEGY -> parameters.get("value.schema.namespace").get
+        )
+
+        val avroDf = dataFrame.select(to_avro(columns, schemaRegistryConfig) as 'value)
+
+        avroDf.createOrReplaceTempView(dataFrameName)
+
+        avroDf.cache()
```
Comment (Doe-Ed) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:32]: test if you can remove

Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala]: you can

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.yotpo.metorikku.exceptions.MetorikkuException
+import org.apache.spark.sql.functions.struct
+import za.co.absa.abris.avro.functions.to_avro
+import za.co.absa.abris.avro.read.confluent.SchemaManager
+
+object ToAvro {
+
+  val message = "You need to send the following parameters to output to Avro format:" +
+    "table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace." +
+    "Will create an entry in the schema registry under <schema.registry.topic>-<value.schema.namespace>.<value.schema_name>"
```
Comment (amitco1) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:12]: ...under a **subject**: <schema.registry.topic<...

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.yotpo.metorikku.exceptions.MetorikkuException
+import org.apache.spark.sql.functions.struct
+import za.co.absa.abris.avro.functions.to_avro
+import za.co.absa.abris.avro.read.confluent.SchemaManager
+
+object ToAvro {
+
+  val message = "You need to send the following parameters to output to Avro format:" +
+    "table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace." +
```
Comment (amitco1) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:11]: delete "." ate the end of value.schema.namespace

File: src/test/scala/com/yotpo/metorikku/code/steps/test/ToAvroTests.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      ToAvro.run(sparkSession, "MetricName", "DataFrameName", params)
+    }
+  }
+
+  test("ToAvro Missing Params") {
+    assertThrows[com.yotpo.metorikku.exceptions.MetorikkuException] {
+      ToAvro.run(sparkSession, "MetricName", "DataFrameName", None)
+    }
+  }
+}
```
Comment (amitco1) [inline src/test/scala/com/yotpo/metorikku/code/steps/test/ToAvroTests.scala:51]: add this description to the readme, after watermark - 

Metorikku supports to_avro() method which turns a dataframe into Avro records and send it to kafka.
the method required parameters are: table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace, kafka.url, schema.registry.topic .
The method will auto create a topic if not exist. also, if not exist, a subject will be created in schema registry. the subject name will be: <schema.registry.topic>-<value.schema.namespace>.<value.schema_name>

ADD AN EXAMPLE :)

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala (hunk 3 above)
Comment (amitco1) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:11]: also add this parameters:
kafka.url, schema.registry.topic

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala (hunk 2 above)
Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:12]: done

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala (hunk 3 above)
Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:11]: done

Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:11]: schema.registry.topic already exists, adding kafka.url

File: src/test/scala/com/yotpo/metorikku/code/steps/test/ToAvroTests.scala (hunk 4 above)
Comment (Doe-Ed) [reply src/test/scala/com/yotpo/metorikku/code/steps/test/ToAvroTests.scala:51]: done

File: README.md [hunk 5]
```diff
@@ -334,6 +334,24 @@ You can use watermarking by adding a new udf step in your metric:
...
+##### ToAvro
+Metorikku supports to_avro() method which turns a dataframe into Avro records and send these to kafka.
+The method requires the following parameters: table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace, schema.registry.topic, and kafka.url.
+The method will auto-create a topic in Kafka if one doesn't exist; moreover, a subject will be created in the schema registry (if one doesn't already exist). The subject name will be: <schema.registry.topic>-<value.schema.namespace>.<value.schema_name>.
+<br/>You can use ToAvro by adding a new udf step in your metric:
+```yaml
+- dataFrameName: dataframe
+  classpath: com.yotpo.metorikku.code.steps.ToAvro
+  params:
+    table: my_table
```
Comment (liebstein) [inline README.md:346]: What exactly does table mean?
We pass the dataframe name. why we need table as well?

Comment (Doe-Ed) [reply README.md:346]: Table is the table we will load and convert to Avro. Dataframe is the name of the dataframe we call createOrReplaceTempView on (with the converted table)

Comment (liebstein) [reply README.md:346]: right. sorry

Comment (Doe-Ed) [reply README.md:346]: That is, we do ss.table("my_table")

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+object ToAvro {
+
+  val message = "You need to send the following parameters to output to Avro format:" +
+    "table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace, kafka.url " +
+    "Will create an entry in the schema registry under: <schema.registry.topic>-<value.schema.namespace>.<value.schema_name>"
+
+  def run(ss: org.apache.spark.sql.SparkSession, metricName: String, dataFrameName: String, params: Option[Map[String, String]]): Unit = {
+    params match {
+      case Some(parameters) => {
+        val table = parameters.get("table").get
```
Comment (liebstein) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:17]: Maybe wrap in try-catch and raise a proper exception when a parameter is missing.

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+          SchemaManager.PARAM_VALUE_SCHEMA_NAMING_STRATEGY               -> SchemaManager.SchemaStorageNamingStrategies.TOPIC_RECORD_NAME,
+          SchemaManager.PARAM_VALUE_SCHEMA_NAME_FOR_RECORD_STRATEGY      -> parameters.get("value.schema.name").get,
+          SchemaManager.PARAM_VALUE_SCHEMA_NAMESPACE_FOR_RECORD_STRATEGY -> parameters.get("value.schema.namespace").get
+        )
+
+        val avroDf = dataFrame.select(to_avro(columns, schemaRegistryConfig) as 'value)
+
+        avroDf.createOrReplaceTempView(dataFrameName)
+
+        val query = avroDf
```
Comment (shirbr) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:33]: need to remove this part - you can use kafka output writer with topic and kafka url

Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala]: done

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala (hunk 6 above)
Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:18]: :\ good idea

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 8]
```diff
@@ -11,20 +11,23 @@ object ToAvro {
...
+  }
+  private val InputMatcher = new InputMatcher("table", "schema.registry.url", "schema.registry.topic", "value.schema.name", "value.schema.namespace")
 
   def run(ss: org.apache.spark.sql.SparkSession, metricName: String, dataFrameName: String, params: Option[Map[String, String]]): Unit = {
-    params match {
-      case Some(parameters) => {
-        val table = parameters.get("table").get
-        val dataFrame = ss.table(table)
+    params.get match {
+      case InputMatcher(tableName, schemaRegistryUrl, schemaRegistryTopic, valueSchemaName, valueSchemaNamespace) => {
```
Comment (liebstein) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:21]: like it very much

File: src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala [hunk 9]
```diff
@@ -11,20 +11,23 @@ object ToAvro {
   val message = "You need to send the following parameters to output to Avro format:" +
     "table, schema.registry.url, schema.registry.topic, value.schema.name, value.schema.namespace " +
     "Will create an entry in the schema registry under: <schema.registry.topic>-<value.schema.namespace>.<value.schema_name>"
+  private class InputMatcher[K](ks: K*) {
+    def unapplySeq[V](m: Map[K, V]): Option[Seq[V]] = if (ks.forall(m.contains)) Some(ks.map(m)) else None
```
Comment (liebstein) [inline src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:15]: Why do you have to implement it?

Comment (Doe-Ed) [reply src/main/scala/com/yotpo/metorikku/code/steps/ToAvro.scala:15]: Will throw Metorikku exception in-case any of the InputMatcher arguments are missing...

**Pull request merged**